app/post_acknowledgements.go

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package app
+
+import (
+	"encoding/json"
+	"errors"
+	"net/http"
+
+	"github.com/mattermost/mattermost-server/v6/app/request"
+	"github.com/mattermost/mattermost-server/v6/model"
+	"github.com/mattermost/mattermost-server/v6/shared/mlog"
+	"github.com/mattermost/mattermost-server/v6/store"
+)
+
+func (a *App) SaveAcknowledgementForPost(c *request.Context, userID, postID string) (*model.PostAcknowledgement, *model.AppError) {
+	post, err := a.GetSinglePost(postID, false)
+	if err != nil {
+		return nil, err
+	}
+
+	channel, err := a.GetChannel(c, post.ChannelId)
+	if err != nil {
+		return nil, err
+	}
+
+	if channel.DeleteAt > 0 {
+		return nil, model.NewAppError("SaveAcknowledgementForPost", "api.acknowledgement.save.archived_channel.app_error", nil, "", http.StatusForbidden)
+	}
+
+	acknowledgedAt := model.GetMillis()
+	acknowledgement, nErr := a.Srv().Store().PostAcknowledgement().Save(userID, postID, acknowledgedAt)
+
+	if nErr != nil {
+		var appErr *model.AppError
+		switch {
+		case errors.As(nErr, &appErr):
+			return nil, appErr
+		default:
+			return nil, model.NewAppError("SaveAcknowledgementForPost", "app.acknowledgement.save.save.app_error", nil, "", http.StatusInternalServerError).Wrap(nErr)
+		}
+	}
+
+	a.Srv().Go(func() {
+		a.sendAcknowledgementEvent(model.WebsocketEventAcknowledgementAdded, acknowledgement, post)
+	})
+
+	return acknowledgement, nil
+}
+
+func (a *App) DeleteAcknowledgementForPost(c *request.Context, userID, postID string) *model.AppError {
+	post, err := a.GetSinglePost(postID, false)
+	if err != nil {
+		return err
+	}
+
+	channel, err := a.GetChannel(c, post.ChannelId)
+	if err != nil {
+		return err
+	}
+
+	if channel.DeleteAt > 0 {
+		return model.NewAppError("DeleteAcknowledgementForPost", "api.acknowledgement.delete.archived_channel.app_error", nil, "", http.StatusForbidden)
+	}
+
+	oldAck, nErr := a.Srv().Store().PostAcknowledgement().Get(userID, postID)
+
+	if nErr != nil {
+		var nfErr *store.ErrNotFound
+		switch {
+		case errors.As(nErr, &nfErr):
+			return model.NewAppError("GetPostAcknowledgement", "app.acknowledgement.get.app_error", nil, "", http.StatusNotFound).Wrap(nErr)
+		default:
+			return model.NewAppError("GetPostAcknowledgement", "app.acknowledgement.get.app_error", nil, "", http.StatusInternalServerError).Wrap(nErr)
+		}
+	}
+
+	if model.GetMillis()-oldAck.AcknowledgedAt > 5*60*1000 {
+		return model.NewAppError("DeleteAcknowledgementForPost", "api.acknowledgement.delete.deadline.app_error", nil, "", http.StatusForbidden)
+	}
+
+	nErr = a.Srv().Store().PostAcknowledgement().Delete(oldAck)
+
+	if nErr != nil {
+		return model.NewAppError("DeleteAcknowledgementForPost", "app.acknowledgement.delete_all_with_emoji_name.get_acknowledgement.app_error", nil, "", http.StatusInternalServerError).Wrap(nErr)
+	}
+
+	oldAck.AcknowledgedAt = 0
```

Comment:
To avoid getting the ack from the store a second time, but yes I can change to have the real value returned to the client.

Reply:
@lieut-data now that I think of it, since we are deleting there is no way to get the deleted ack back, how can we avoid the race?

Reply:
I haven't seen the client-side code, but thinking out loud: since the ack's primary key is deterministic, I imagine that a `POST /ack` followed by a `DELETE /ack` followed by a `POST /ack` might result in some clients getting `ack, ack, delete` (e.g. out of order websockets). Since the only unique thing about each is the `AcknowledgedAt`, the client can just ignore anything with value strictly less than the currently known timestamp, keeping the second `ack` in this case and ignoring the `delete` to avoid the race.

Reply:
oh I see, ok I'll add the necessary changes in the webapp PR as well